### PR TITLE
nullary and ternary expressions

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9c1f6281939469b7e40735dd038f325d98413648).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85611876

File: src/util/std_expr.h [hunk 1]
```diff
@@ -18,15 +18,67 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "mathematical_types.h"
 #include "std_types.h"
 
+/// An expression without operands
+class nullary_exprt:public exprt
```
Comment (smowton) [inline src/util/std_expr.h]: Nitpick here and elsewhere: spacing: `...exprt : public exprt`

**Review by smowton — CHANGES_REQUESTED** [review]

**Review by smowton — APPROVED** [review]
oops

File: src/util/std_expr.h [hunk 2]
```diff
@@ -18,6 +18,34 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  DEPRECATED("use nullary_exprt(id, type) instead")
+  explicit nullary_exprt(const irep_idt &_id) : exprt(_id)
+  {
+  }
+
+  nullary_exprt(const irep_idt &_id, const typet &_type) : exprt(_id, _type)
+  {
+  }
+
+  /// remove all operand methods
```
Comment (tautschnig) [inline src/util/std_expr.h:35]: What about copy_to_operands, move_to_operands?

File: src/util/std_expr.h [hunk 3]
```diff
@@ -207,14 +234,13 @@ inline void validate_expr(const symbol_exprt &value)
 
 
 /// \brief Expression to hold a nondeterministic choice
-class nondet_symbol_exprt:public exprt
+class nondet_symbol_exprt : public nullary_exprt
 {
 public:
   /// \param identifier: Name of symbol
   /// \param : type Type of symbol
```
Comment (tautschnig) [inline src/util/std_expr.h]: The `:` is in the wrong place, tripping up Doxygen. Not your fault, but nowadays it stops CI...

File: src/util/string_constant.cpp [hunk 4]
```diff
@@ -12,14 +12,13 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "c_types.h"
 #include "std_expr.h"
 
-string_constantt::string_constantt():
-  exprt(ID_string_constant)
+string_constantt::string_constantt() : nullary_exprt(ID_string_constant)
```
Comment (tautschnig) [inline src/util/string_constant.cpp:15]: Should this one also be deprecated?

**Review by tautschnig — APPROVED** [review]
Looks good modulo the below (partly required - to make CI happy) comments.

File: src/util/std_expr.h (hunk 2 above)
Comment (kroening) [reply src/util/std_expr.h:35]: yes, can go away as well

File: src/util/string_constant.cpp (hunk 4 above)
Comment (kroening) [reply src/util/string_constant.cpp:15]: yes, while at it

File: src/util/std_expr.h (hunk 3 above)
Comment (kroening) [reply src/util/std_expr.h]: fixed

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: dfb1c9575f6f5d3b2f7c96ff56f98407bfb24ff7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85638395

**Pull request merged**